File: ios/Sasquatsh/Services/PlanningService.swift
import Foundation

protocol PlanningServiceProtocol: Sendable {
    func getGroupSessions(groupId: String) async throws -> [PlanningSession]
    func getMySessions() async throws -> [PlanningSession]
    func getSession(id: String) async throws -> PlanningSession
    func createSession(input: CreatePlanningSessionInput) async throws -> PlanningSession
    func submitResponse(sessionId: String, input: PlanningResponseInput) async throws
    func suggestGame(sessionId: String, input: SuggestGameInput) async throws -> GameSuggestion
    func voteForGame(sessionId: String, suggestionId: String) async throws
    func unvoteGame(sessionId: String, suggestionId: String) async throws
    func removeSuggestion(sessionId: String, suggestionId: String) async throws
    func finalizeSession(sessionId: String, selectedDateId: String?, selectedGameId: String?) async throws -> FinalizeResponse
    func cancelSession(sessionId: String) async throws
    func deleteSession(sessionId: String) async throws
    func addItem(sessionId: String, input: AddPlanningItemInput) async throws -> PlanningItem
    func claimItem(sessionId: String, itemId: String) async throws
    func unclaimItem(sessionId: String, itemId: String) async throws
    func removeItem(sessionId: String, itemId: String) async throws
    func addInvitees(sessionId: String, userIds: [String]) async throws
    func updateSettings(sessionId: String, tableCount: Int?) async throws
    func scheduleSessions(sessionId: String, schedule: [ScheduleEntry]) async throws
}

struct FinalizeResponse: Codable {
    let eventId: String
    let message: String
}

struct ScheduleEntry: Codable {
    let suggestionId: String
    let tableNumber: Int
    let slotIndex: Int
}

final class PlanningService: PlanningServiceProtocol {
    private let api: APIClient

    init(api: APIClient) {
        self.api = api
    }

    func getGroupSessions(groupId: String) async throws -> [PlanningSession] {
        try await api.get("planning", queryItems: [.init(name: "groupId", value: groupId)], authenticated: true)
    }

    func getMySessions() async throws -> [PlanningSession] {
        try await api.get("planning", queryItems: [.init(name: "mine", value: "true")], authenticated: true)
    }

    func getSession(id: String) async throws -> PlanningSession {
        try await api.get("planning", queryItems: [.init(name: "id", value: id)], authenticated: true)
    }

    func createSession(input: CreatePlanningSessionInput) async throws -> PlanningSession {
        try await api.post("planning", body: input)
    }

    func submitResponse(sessionId: String, input: PlanningResponseInput) async throws {
        try await api.postVoid("planning", body: input, queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "respond")
        ])
    }

    func suggestGame(sessionId: String, input: SuggestGameInput) async throws -> GameSuggestion {
        try await api.post("planning", body: input, queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "suggest-game")
        ])
    }

    func voteForGame(sessionId: String, suggestionId: String) async throws {
        try await api.postVoid("planning", queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "vote-game"),
            .init(name: "suggestionId", value: suggestionId)
        ])
    }

    func unvoteGame(sessionId: String, suggestionId: String) async throws {
        try await api.postVoid("planning", queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "unvote-game"),
            .init(name: "suggestionId", value: suggestionId)
        ])
    }

    func removeSuggestion(sessionId: String, suggestionId: String) async throws {
        try await api.postVoid("planning", queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "remove-suggestion"),
            .init(name: "suggestionId", value: suggestionId)
        ])
    }

    func finalizeSession(sessionId: String, selectedDateId: String?, selectedGameId: String?) async throws -> FinalizeResponse {
        struct FinalizeInput: Codable {
            let selectedDateId: String?
            let selectedGameId: String?
        }
        return try await api.put("planning", body: FinalizeInput(selectedDateId: selectedDateId, selectedGameId: selectedGameId), queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "finalize")
        ])
    }

    func cancelSession(sessionId: String) async throws {
        try await api.putVoid("planning", queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "cancel")
        ])
    }

    func deleteSession(sessionId: String) async throws {
        try await api.deleteVoid("planning", queryItems: [.init(name: "id", value: sessionId)])
    }

    func addItem(sessionId: String, input: AddPlanningItemInput) async throws -> PlanningItem {
        try await api.post("planning", body: input, queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "add-item")
        ])
    }

    func claimItem(sessionId: String, itemId: String) async throws {
        try await api.postVoid("planning", queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "claim-item"),
            .init(name: "itemId", value: itemId)
        ])
    }

    func unclaimItem(sessionId: String, itemId: String) async throws {
        try await api.postVoid("planning", queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "unclaim-item"),
            .init(name: "itemId", value: itemId)
        ])
    }

    func removeItem(sessionId: String, itemId: String) async throws {
        try await api.postVoid("planning", queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "remove-item"),
            .init(name: "itemId", value: itemId)
        ])
    }

    func addInvitees(sessionId: String, userIds: [String]) async throws {
        try await api.postVoid("planning", body: ["userIds": userIds], queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "add-invitees")
        ])
    }

    func updateSettings(sessionId: String, tableCount: Int?) async throws {
        struct Body: Encodable { let tableCount: Int? }
        let _: UpdateSettingsResponse = try await api.post("planning", body: Body(tableCount: tableCount), queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "update-settings")
        ])
    }

    func scheduleSessions(sessionId: String, schedule: [ScheduleEntry]) async throws {
        struct Body: Encodable { let schedule: [ScheduleEntry] }
        let _: ScheduleResponse = try await api.post("planning", body: Body(schedule: schedule), queryItems: [
            .init(name: "id", value: sessionId),
            .init(name: "action", value: "schedule-sessions")
        ])
    }
}

private struct UpdateSettingsResponse: Decodable {
    let message: String
    let tableCount: Int?
}

private struct ScheduleResponse: Decodable {
    let message: String
}

File: android/app/src/main/java/com/sasquatsh/app/services/PlanningService.kt
package com.sasquatsh.app.services

import com.sasquatsh.app.models.*
import com.sasquatsh.app.services.api.PlanningApi
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import javax.inject.Inject
import javax.inject.Singleton

data class FinalizeResult(
    val message: String,
    val eventId: String
)

@Singleton
class PlanningService @Inject constructor(
    private val planningApi: PlanningApi,
    private val moshi: Moshi
) {

    private val sessionListType = Types.newParameterizedType(List::class.java, PlanningSession::class.java)

    suspend fun getGroupSessions(groupId: String): List<PlanningSession> {
        val response = planningApi.getGroupSessions(groupId)
        if (!response.isSuccessful) throw Exception("Failed to load planning sessions")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter<List<PlanningSession>>(sessionListType).fromJson(json) ?: emptyList()
    }

    suspend fun getMySessions(): List<PlanningSession> {
        val response = planningApi.getMySessions()
        if (!response.isSuccessful) throw Exception("Failed to load my sessions")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter<List<PlanningSession>>(sessionListType).fromJson(json) ?: emptyList()
    }

    suspend fun getSession(id: String): PlanningSession {
        val response = planningApi.getSession(id)
        if (!response.isSuccessful) throw Exception("Failed to load session")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter(PlanningSession::class.java).fromJson(json)
            ?: throw Exception("Failed to parse session")
    }

    suspend fun createSession(input: CreatePlanningSessionInput): PlanningSession {
        val response = planningApi.createSession(input)
        if (!response.isSuccessful) throw Exception("Failed to create session")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter(PlanningSession::class.java).fromJson(json)
            ?: throw Exception("Failed to parse session")
    }

    suspend fun submitResponse(sessionId: String, input: PlanningResponseInput) {
        val response = planningApi.submitResponse(sessionId = sessionId, input = input)
        if (!response.isSuccessful) throw Exception("Failed to submit response")
    }

    suspend fun suggestGame(sessionId: String, input: SuggestGameInput): GameSuggestion {
        val response = planningApi.suggestGame(sessionId = sessionId, input = input)
        if (!response.isSuccessful) throw Exception("Failed to suggest game")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter(GameSuggestion::class.java).fromJson(json)
            ?: throw Exception("Failed to parse suggestion")
    }

    suspend fun voteForGame(sessionId: String, suggestionId: String) {
        val response = planningApi.voteForGame(sessionId = sessionId, suggestionId = suggestionId)
        if (!response.isSuccessful) throw Exception("Failed to vote for game")
    }

    suspend fun unvoteGame(sessionId: String, suggestionId: String) {
        val response = planningApi.unvoteGame(sessionId = sessionId, suggestionId = suggestionId)
        if (!response.isSuccessful) throw Exception("Failed to remove vote")
    }

    suspend fun removeSuggestion(sessionId: String, suggestionId: String) {
        val response = planningApi.removeSuggestion(sessionId = sessionId, suggestionId = suggestionId)
        if (!response.isSuccessful) throw Exception("Failed to remove suggestion")
    }

    suspend fun finalizeSession(sessionId: String, selectedDateId: String?, selectedGameId: String?): FinalizeResult {
        val input = mutableMapOf<String, Any>()
        selectedDateId?.let { input["selectedDateId"] = it }
        selectedGameId?.let { input["selectedGameId"] = it }
        val response = planningApi.finalizeSession(sessionId = sessionId, input = input)
        if (!response.isSuccessful) throw Exception("Failed to finalize session")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        val map = moshi.adapter(Map::class.java).fromJson(json)
        return FinalizeResult(
            message = map?.get("message") as? String ?: "Session finalized",
            eventId = map?.get("eventId") as? String ?: ""
        )
    }

    suspend fun cancelSession(sessionId: String) {
        val response = planningApi.cancelSession(sessionId = sessionId)
        if (!response.isSuccessful) throw Exception("Failed to cancel session")
    }

    suspend fun addItem(sessionId: String, input: AddPlanningItemInput): PlanningItem {
        val response = planningApi.addItem(sessionId = sessionId, input = input)
        if (!response.isSuccessful) throw Exception("Failed to add item")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter(PlanningItem::class.java).fromJson(json)
            ?: throw Exception("Failed to parse item")
    }

    suspend fun claimItem(sessionId: String, itemId: String) {
        val response = planningApi.claimItem(sessionId = sessionId, itemId = itemId)
        if (!response.isSuccessful) throw Exception("Failed to claim item")
    }

    suspend fun unclaimItem(sessionId: String, itemId: String) {
        val response = planningApi.unclaimItem(sessionId = sessionId, itemId = itemId)
        if (!response.isSuccessful) throw Exception("Failed to unclaim item")
    }

    suspend fun removeItem(sessionId: String, itemId: String) {
        val response = planningApi.removeItem(sessionId = sessionId, itemId = itemId)
        if (!response.isSuccessful) throw Exception("Failed to remove item")
    }

    suspend fun scheduleSessions(sessionId: String, schedule: List<ScheduleEntry>) {
        val body = mapOf("schedule" to schedule)
        val response = planningApi.scheduleSessions(sessionId = sessionId, body = body)
        if (!response.isSuccessful) throw Exception("Failed to schedule sessions")
    }

    suspend fun updateSettings(sessionId: String, tableCount: Int?) {
        val body = mutableMapOf<String, Any?>()
        body["tableCount"] = tableCount
        val response = planningApi.updateSettings(sessionId = sessionId, body = body)
        if (!response.isSuccessful) throw Exception("Failed to update settings")
    }
}
